### PR TITLE
Callbacks [wip]

### PR DESCRIPTION
**Pull request by hadim**
This PR is based on https://github.com/PyTorchLightning/pytorch-lightning/pull/849.

A new callback mechanism is added where a user can pass a list of callbacks to the trainer that is called during the various steps of the trainer's life.

```python
from pytorch_lightning.callbacks import Callback

class PrintCallback(Callback):
    def on_fit_begin(self):
        print("Training is started!")

    def on_fit_end(self):
        print(f"Training is done. The logs are: {self.trainer.logs}")

# a list of callbacks
callbacks = [PrintCallback()]
trainer = Trainer(callbacks=callbacks)
```

It allows inserting custom logic in the training loop outside of the `LightningModule`.

Let me know what do you guys think.

### CODE REVIEW

File: pytorch_lightning/trainer/training_loop.py [hunk 1]
```diff
@@ -305,6 +306,11 @@ def process_output(self, output, train):
         pass
 
     def train(self):
+
+        # Train begin callbacks
+        for callback in self.callbacks:
+            callback.on_train_begin()
```
Comment (kuynzereb) [inline pytorch_lightning/trainer/training_loop.py]: I would suggest to add something like `Trainer` hooks. So we can just call `self.on_train_begin()` in this place, and move
```
for callback in self.callbacks:
    callback.on_train_begin()
```
into `trainer.on_train_begin()`. I think it would be more clear because actually there are other things which should be called at certain points in training (for example, model hooks)

File: environment.yml [hunk 2]
```diff
@@ -0,0 +1,35 @@
+# Usage: `conda env update -f environment.yml`
```
Comment (Borda) [inline environment.yml]: what is the reason for adding `conda` env here?

Comment (hadim) [reply environment.yml]: don't pay attention to this, it's because this PR is based on #849

File: pytorch_lightning/trainer/training_loop.py (hunk 1 above)
Comment (hadim) [reply pytorch_lightning/trainer/training_loop.py]: Agree. The logic now lives in `TrainerCallbackHookMixin`.

File: environment.yml (hunk 2 above)
Comment (Borda) [reply environment.yml]: good to know lol :]

File: pytorch_lightning/callbacks/base.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    def __init__(self):
+        self._trainer = None
+
+    @property
+    def trainer(self):
+        assert self._trainer is not None, _NO_TRAINER_ERROR_MSG
+        return self._trainer
+
+    @property
+    def default_save_path(self):
```
Comment (Borda) [inline pytorch_lightning/callbacks/base.py]: do we need this method which just wraps one more call?

File: pytorch_lightning/callbacks/base.py [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+        """Trainer default save path.
+        """
+        return self.trainer.default_save_path
```
Comment (Borda) [inline pytorch_lightning/callbacks/base.py]:
```suggestion
        return self._trainer.default_save_path
```

File: pytorch_lightning/callbacks/base.py (hunk 3 above)
Comment (hadim) [reply pytorch_lightning/callbacks/base.py]: No, we don't.

My intention here was to encourage callback's creators to use this instead of using a `dir_path` argument in the init of the callback so all the files stay in the trainer's `default_save_path`. 

I can remove it if you want.

File: pytorch_lightning/trainer/trainer.py [hunk 5]
```diff
@@ -168,6 +172,18 @@ def __init__(
 
                     trainer = Trainer(early_stop_callback=early_stop_callback)
 
+            callback (:class:`.Callback`): Add a list of callbacks.
+                Example::
+                    from pytorch_lightning.callbacks import Callback
+                    class PrintCallback(Callback):
```
Comment (williamFalcon) [inline pytorch_lightning/trainer/trainer.py]: callback should be passed in trainer and the model no?

Comment (williamFalcon) [reply pytorch_lightning/trainer/trainer.py]: i mean, each call should have those args.

Comment (hadim) [reply pytorch_lightning/trainer/trainer.py]: Don't you prefer to have those variables as a callback property such as:

```python 
class Callback(abc.ABC):
    """Abstract base class used to build new callbacks."""

    def __init__(self):
        self._trainer = None

    @property
    def trainer(self):
        assert self._trainer is not None, _NO_TRAINER_ERROR_MSG
        return self._trainer

    @property
    def module(self):
        return self._trainer.get_model()

    @property
    def default_save_path(self):
        """Trainer default save path.
        """
        return self._trainer.default_save_path

    @property
    def rank(self):
        """Current trainer rank.
        """
        return self._trainer.proc_rank
```

It makes the code in `pl.Trainer` easier to read and update if needed. and overall make the callback class less verbose (especially since all callbacks have the same arguments).

**Comment by hadim** [comment]
@williamFalcon about immutability, I had a look and I don't think this is something easy todo in Python without heavy hacks. See https://discuss.python.org/t/immutability-in-python-is-really-hard/2536/2

I think we should just make it very clear in the doc/docstring that callbacks MUST NOT modify the pl module and the trainer.

File: pytorch_lightning/trainer/trainer.py (hunk 5 above)
Comment (williamFalcon) [reply pytorch_lightning/trainer/trainer.py]: i’m not sure i agree. 

i thought about having a self.trainer and self.model but that means the user has to read the docs for this. whereas if you pass it in as an arg, it’s clear what’s available. 

but i see the value in both, so could go either way

**Comment by williamFalcon** [comment]
why do we want to add this limitation? shouldn’t the user be ok with messing around with the trainer or model if they want to in a callback?

**Comment by williamFalcon** [comment]
also, should we rethink the name “Callback”?

this ALWAYS ends up needing a lengthy explanation about what it is. why not call it something more intuitive? 

Maybe something like:
“FunctionBlock”
“CodeBlock”
“Snippet”

just brainstorming here lol. i just think it’s super confusing for newcomers and non engineers.

File: pytorch_lightning/trainer/trainer.py (hunk 5 above)
Comment (hadim) [reply pytorch_lightning/trainer/trainer.py]: Let's wait for other comments. I don't have a strong feeling about it but I find the property pattern more elegant.

As for the doc, we should except the user to read it before using callbacks. This is not an extremely complicated pattern neither with plenty of parameters, so I except it should be fine.

Comment (williamFalcon) [reply pytorch_lightning/trainer/trainer.py]: i disagree. “expecting” the user to read the docs means the api is not intuitive or obvious.

Comment (ethanwharris) [reply pytorch_lightning/trainer/trainer.py]: I agree with @williamFalcon on this, arguments to the mthods seems much more intuitive

Comment (hadim) [reply pytorch_lightning/trainer/trainer.py]: Done.

**Comment by hadim** [comment]
I haven't totally incorporated the 3 already existant callbacks in the new system (they still use the new `Callback` class). I think the merge of the existant callbacks should be done in another PR to not make this one too large in terms of changes.

About the name, I like `Callback` but I don't have strong feelings about it so whatever you decide is good for me.

**Comment by jeremyjordan** [comment]
> I haven't totally incorporated the 3 already existant callbacks in the new system (they still use the new Callback class). I think the merge of the existant callbacks should be done in another PR to not make this one too large in terms of changes.

Regarding this, should we update the `early_stopping` and `model_checkpointing` args to serve as simple flags to include the callbacks (or whatever we call them) in the list?

```
Trainer(..., checkpoint_callback=True)  # use default
Trainer(..., checkpoint_callback=False)  # disable

checkpoint_callback = ModelCheckpoint(filepath='my_path')
Trainer(..., checkpoint_callback=checkpoint_callback)  # warn user to pass via callbacks arg
Trainer(..., callbacks=[checkpoint_callback])  # use checkpoint callback with user provided spec
```

**Comment by jeremyjordan** [comment]
> why do we want to add this limitation? shouldn’t the user be ok with messing around with the trainer or model if they want to in a callback?

Yes, but if the user can modify the trainer then the callback quickly transitions from non-essential to essential for reproducibility. 

However, there are cases when it would be desirable to modify objects in the trainer. For example, fast.ai has a callback for a [learning rate finder](https://docs.fast.ai/callbacks.html#LRFinder) which is quite useful. This would be something that the user would want to enable when first building a model but then disable once they find a good range and want to start training.

**Comment by Borda** [comment]
> Regarding this, should we update the early_stopping and model_checkpointing args to serve as simple flags to include the callbacks (or whatever we call them) in the list?

Then we shall check that they are not duplicated, meaning when they are added via simple flag and later among callbacks?

**Comment by jeremyjordan** [comment]
Ah yes, good point 😄 Perhaps in that case, we notice that an instance of `ModelCheckpoint` was provided in the callback list and issue a warning that the user provided instance will be used in place of the default instance. 

```
checkpoint_callback = ModelCheckpoint(filepath='my_path')

Trainer(..., checkpoint_callback=True)  # use default
Trainer(..., callbacks=[checkpoint_callback])  # use checkpoint callback with user provided spec
Trainer(..., checkpoint_callback=True, callbacks=[checkpoint_callback])  # use checkpoint callback with user provided spec and alert user via a warning
```

What do you think? Does this resolve any potential confusion?

**Comment by Borda** [comment]
I think that the question is very hard because you do not have much clues if double checkpoint is mistake or desired state... Maybe you would need to match also parameters of the ChP to determine whether or not they are identical...

**Comment by hadim** [comment]
Maybe a warning is enough? But then if double checkpoint if often used by someone a warning each time could become annoying...

**Comment by jeremyjordan** [comment]
I don't think that having two checkpoint callbacks would be a common use case (eg. it's not even possible today). If the callback is constructed in both places, I'd lean towards telling the user that Lightning is going to use the one that they provided. If they absolutely required having two instances of the same callback, then they can pass both instances in the callback list.

**Comment by hadim** [comment]
What do you think if we finish the design of the callback system and discuss the way we integrate the existing callbacks in a new PR/issue?

Do you have any comments about the one implemented in this PR?

File: pytorch_lightning/trainer/trainer.py [hunk 6]
```diff
@@ -62,13 +64,15 @@ class Trainer(TrainerIOMixin,
...
               TrainerCallbackConfigMixin,
+              TrainerCallbackHookMixin
               ):
 
     def __init__(
             self,
             logger: Union[LightningLoggerBase, bool] = True,
             checkpoint_callback: Union[ModelCheckpoint, bool] = True,
             early_stop_callback: Optional[Union[EarlyStopping, bool]] = None,
+            callbacks: list = [],
```
Comment (awaelchli) [inline pytorch_lightning/trainer/trainer.py]: I recently added typehints to the Trainer and would like to suggest to make the type List[Callback]

File: pytorch_lightning/trainer/trainer.py [hunk 7]
```diff
@@ -168,6 +172,18 @@ def __init__(
 
                     trainer = Trainer(early_stop_callback=early_stop_callback)
 
+            callback (:class:`.Callback`): Add a list of callbacks.
```
Comment (awaelchli) [inline pytorch_lightning/trainer/trainer.py]: The :class annotation can be omitted because sphinx will autogenerate one given the typehint above

Comment (awaelchli) [reply pytorch_lightning/trainer/trainer.py]: I also just saw right now that the documentation refers to "callback" but the actual argument is plural, callbacks.

File: pytorch_lightning/callbacks/base.py [hunk 8]
```diff
@@ -8,61 +8,61 @@
 import abc
 
 
-_NO_TRAINER_ERROR_MSG = ".set_trainer() should be called after the callback initialization"
-
-
 class Callback(abc.ABC):
```
[the remark this answers is not in the document]
Comment (hadim) [reply pytorch_lightning/callbacks/base.py]: Definitively.

File: pytorch_lightning/callbacks/early_stopping.py [hunk 9]
```diff
@@ -64,7 +64,7 @@ def __init__(self, monitor: str = 'val_loss', min_delta: float = 0.0, patience:
         self.monitor_op = mode_dict[mode]
         self.min_delta *= 1 if self.monitor_op == np.greater else -1
 
-        self.on_train_begin()
+        self.on_train_begin(None, None)
```
[the remark this answers is not in the document]
Comment (hadim) [reply pytorch_lightning/callbacks/early_stopping.py]: I agree it's weird and should be fixed.

File: pytorch_lightning/callbacks/base.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (jeremyjordan) [reply pytorch_lightning/callbacks/base.py]: awesome! i’d prefer that we standardize to the `_start` suffix since that would be the least impactful to current users using model hooks already

**Comment by jeremyjordan** [comment]
@hadim could we consolidate the code such that model hooks and callbacks are invoked at the same time, and within the context of the a profiler?

```
with self.profiler.profile('on_epoch_start'):
    self.on_epoch_start()  # callbacks
    if self.is_function_implemented('on_epoch_start'):
        model = self.get_model()
        model.on_epoch_start()  # model hook
```

**Comment by hadim** [comment]
Yes, we should. Not sure when I'll have time to work on this PR this week. So we could merge it as it is and improve it in other PRs or if someone wants to push some commits here, feel free (I can add you to my fork).

**Comment by williamFalcon** [comment]
@hadim why don't we merge this asap so we can get it in the release.
@jeremyjordan maybe open a new PR to implement your suggestions? i do like the profiler addition here.

Great job everyone!

**Comment by hadim** [comment]
I agree @williamFalcon 

Will rebase soon.

**Comment by hadim** [comment]
Rebasing wasn't a very funny moment here!

Please merge whenever you can @williamFalcon!

I'll open a new issue for what's next.

**Comment by jeremyjordan** [comment]
happy to pick up the work on a separate PR

**Comment by hadim** [comment]
@jeremyjordan look at https://github.com/PyTorchLightning/pytorch-lightning/issues/947

**Comment by Borda** [comment]
@hadim @jeremyjordan GREAT job! Pls could you update also the change log...
